test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -735,3 +735,284 @@ UCS_TEST_P(test_ucp_sockaddr_with_rma_atomic, wireup) {
 }
 
 UCP_INSTANTIATE_ALL_TEST_CASE(test_ucp_sockaddr_with_rma_atomic)
+
+
+class test_ucp_sockaddr_protocols : public test_ucp_sockaddr {
+public:
+    static ucp_params_t get_ctx_params() {
+        ucp_params_t params = test_ucp_sockaddr::get_ctx_params();
+        params.field_mask  |= UCP_PARAM_FIELD_FEATURES;
+        params.features    |= UCP_FEATURE_RMA;
+        /* Atomics not supported for now because need to emulate the case
+         * of using different device than the one selected by default on the
+         * worker for atomic operations */
+        return params;
+    }
+
+    static std::vector<ucp_test_param>
+    enum_test_params(const ucp_params_t& ctx_params,
+                     const std::string& name,
+                     const std::string& test_case_name,
+                     const std::string& tls)
+    {
+        std::vector<ucp_test_param> result;
+        enum_test_params_with_modifier(ctx_params, name, test_case_name, tls,
+                                       result, TEST_MODIFIER_CM);
+        return result;
+    }
+
+    virtual void init() {
+        test_ucp_sockaddr::init();
+        start_listener(cb_type());
+        client_ep_connect();
+    }
+
+    void get_nb(std::string& send_buf, std::string& recv_buf, ucp_rkey_h rkey,
+                std::vector<void*>& reqs)
+    {
+         reqs.push_back(ucp_get_nb(sender().ep(), &send_buf[0], send_buf.size(),
+                                   (uintptr_t)&recv_buf[0], rkey, scomplete_cb));
+    }
+
+    void put_nb(std::string& send_buf, std::string& recv_buf, ucp_rkey_h rkey,
+                std::vector<void*>& reqs)
+    {
+        reqs.push_back(ucp_put_nb(sender().ep(), &send_buf[0], send_buf.size(),
+                                  (uintptr_t)&recv_buf[0], rkey, scomplete_cb));
+        reqs.push_back(ucp_ep_flush_nb(sender().ep(), 0, scomplete_cb));
+    }
+
+protected:
+    typedef void (test_ucp_sockaddr_protocols::*rma_nb_func_t)(
+                    std::string&, std::string&, ucp_rkey_h, std::vector<void*>&);
+
+    void compare_buffers(std::string& send_buf, std::string& recv_buf)
+    {
+        EXPECT_TRUE(send_buf == recv_buf)
+            << "send_buf: '" << send_buf.substr(0, 20) << "..."
+                             << send_buf.substr(send_buf.length() - 20) << "', "
+            << "recv_buf: '" << recv_buf.substr(0, 20) << "..."
+                             << recv_buf.substr(recv_buf.length() - 20) << "'";
```

Comment:
looks not safety, but it's minor since this is a test failure debug msg

Reply:
substr does not change the original string, it's const: http://www.cplusplus.com/reference/string/string/substr/

Reply:
my concern was about this: `recv_buf.substr(recv_buf.length() - 20)`
http://www.cplusplus.com/reference/string/string/substr
"If this (1st arg) is greater than the string length, it throws `out_of_range`."